### ISSUE TITLE
Does 'count_tensor.masked_fill_(count_tensor > 1, 0)' really work in 'select_topk_candidates'?

### ISSUE THREAD

**Issue by INSYangCL** [ultralytics/ultralytics#24574]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to figure out the process of TaskAlignedAssigner. Andin function _select_topk_candidates_, _count_tensor.masked_fill_(count_tensor > 1, 0)_ has been used for filtering invalid bboxes. But considerring to the following code, _count_tensor_ seems to be bigger than 1 only when  repeated indices occur in _topk_idxs[:, :, k : k + 1]_.
```      
for k in range(self.topk):
            # Expand topk_idxs for each value of k and add 1 at the specified positions
            count_tensor.scatter_add_(-1, topk_idxs[:, :, k : k + 1], ones)
```
But it seems that _torch.topk_ never return repeated indices.
```
topk_metrics, topk_idxs = torch.topk(metrics, self.topk, dim=-1, largest=largest)
```
    def select_topk_candidates(self, metrics, largest=True, topk_mask=None):
        """
        Select the top-k candidates based on the given metrics.

        Args:
            metrics (torch.Tensor): A tensor of shape (b, max_num_obj, h*w), where b is the batch size,
                              max_num_obj is the maximum number of objects, and h*w represents the
                              total number of anchor points.
            largest (bool): If True, select the largest values; otherwise, select the smallest values.
            topk_mask (torch.Tensor): An optional boolean tensor of shape (b, max_num_obj, topk), where
                                topk is the number of top candidates to consider. If not provided,
                                the top-k values are automatically computed based on the given metrics.

        Returns:
            (torch.Tensor): A tensor of shape (b, max_num_obj, h*w) containing the selected top-k candidates.
        """
        # (b, max_num_obj, topk)
        topk_metrics, topk_idxs = torch.topk(metrics, self.topk, dim=-1, largest=largest)
        if topk_mask is None:
            topk_mask = (topk_metrics.max(-1, keepdim=True)[0] > self.eps).expand_as(topk_idxs)
        # (b, max_num_obj, topk)
        topk_idxs.masked_fill_(~topk_mask, 0)

        # (b, max_num_obj, topk, h*w) -> (b, max_num_obj, h*w)
        count_tensor = torch.zeros(metrics.shape, dtype=torch.int8, device=topk_idxs.device)
        ones = torch.ones_like(topk_idxs[:, :, :1], dtype=torch.int8, device=topk_idxs.device)
        for k in range(self.topk):
            # Expand topk_idxs for each value of k and add 1 at the specified positions
            count_tensor.scatter_add_(-1, topk_idxs[:, :, k : k + 1], ones)
        # Filter invalid bboxes
        print("count_tensor max:", count_tensor.max())
        count_tensor.masked_fill_(count_tensor > 1, 0)

        return count_tensor.bool()

**Comment by UltralyticsAssistant** [comment]
👋 Hello @INSYangCL, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Since this is a code logic/question report, please provide as much detail as possible to help us review it efficiently 🔍. If you believe this may be a 🐛 bug, please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) showing:
-Your installed `ultralytics` version
-Your PyTorch version
-The exact file/function path you are inspecting
-A small runnable snippet that reproduces the behavior you observed
-Any printed outputs such as `count_tensor max`

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by INSYangCL** [comment]
I have tried _select_topk_candidates_ with _metrics_  as full one tensor, and result in _count_tensor.max() = 1_

**Comment by raimbekovm** [comment]
you're right that `torch.topk` never returns duplicate indices, so within a valid gt row the count is always 1. the `count_tensor > 1` case doesn't come from topk, it comes from the line above the loop:

```python
topk_idxs.masked_fill_(~topk_mask, 0)
```

this forces every invalid topk slot to index 0. a slot is invalid for padding gts (`mask_gt == 0`) or, in the `topk_mask=None` branch, for rows whose max metric is `<= eps`. when a gt row is invalid all `topk` slots collapse to index 0, so `count_tensor[..., 0]` accumulates to `topk` (> 1) and gets zeroed. without it, anchor 0 would be falsely assigned as a positive to every padding/empty gt.

you get `max() == 1` because you passed a full metrics tensor with no `topk_mask`: the default `topk_mask = topk_metrics.max(-1) > eps` is then True everywhere, nothing collapses to 0, and all topk indices are distinct. pass `topk_mask=mask_gt.expand(-1, -1, self.topk)` like `get_pos_mask` does at `tal.py:166`, or include an all-zero row, and you'll see `count_tensor[..., 0]` go above 1 before being filtered.

**Comment by INSYangCL** [comment]
#

> you're right that `torch.topk` never returns duplicate indices, so within a valid gt row the count is always 1. the `count_tensor > 1` case doesn't come from topk, it comes from the line above the loop:
> 
> topk_idxs.masked_fill_(~topk_mask, 0)
> this forces every invalid topk slot to index 0. a slot is invalid for padding gts (`mask_gt == 0`) or, in the `topk_mask=None` branch, for rows whose max metric is `<= eps`. when a gt row is invalid all `topk` slots collapse to index 0, so `count_tensor[..., 0]` accumulates to `topk` (> 1) and gets zeroed. without it, anchor 0 would be falsely assigned as a positive to every padding/empty gt.
> 
> you get `max() == 1` because you passed a full metrics tensor with no `topk_mask`: the default `topk_mask = topk_metrics.max(-1) > eps` is then True everywhere, nothing collapses to 0, and all topk indices are distinct. pass `topk_mask=mask_gt.expand(-1, -1, self.topk)` like `get_pos_mask` does at `tal.py:166`, or include an all-zero row, and you'll see `count_tensor[..., 0]` go above 1 before being filtered.

@raimbekovm I see. Thank you for your answer! BTW, I wonder whether the whole zero row in ```topk_idxs``` caused by ```topk_mask``` is the only case supporting the necessity of ```count_tensor.masked_fill_(count_tensor > 1, 0)```. Are there any other cases that make duplicate indices need to be filterred?

**Comment by raimbekovm** [comment]
no other case. `torch.topk` returns distinct positions even with fully tied values (verified on an all-equal input: zero duplicate rows), so a valid row never produces a duplicate index. the `masked_fill_(~topk_mask, 0)` collapse to index 0 is the only source. and since `topk_mask` is row-uniform in both paths (real path `mask_gt.expand(...)`, default branch `topk_metrics.max(-1, keepdim) > eps` expanded), a gt row is either fully valid or fully invalid, so the duplicates only ever land on index 0. stress-testing 30 random forward passes, `count > 1` appeared exclusively at column 0 and only on fully-invalid rows.

**Comment by INSYangCL** [comment]
Thank you so much!



---Original---
From: "Murat ***@***.***&gt;
Date: Fri, May 22, 2026 17:28 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] Does 'count_tensor.masked_fill_(count_tensor &gt; 1, 0)' really work in 'select_topk_candidates'? (Issue #24574)


raimbekovm left a comment (ultralytics/ultralytics#24574)
 
no other case. torch.topk returns distinct positions even with fully tied values (verified on an all-equal input: zero duplicate rows), so a valid row never produces a duplicate index. the masked_fill_(~topk_mask, 0) collapse to index 0 is the only source. and since topk_mask is row-uniform in both paths (real path mask_gt.expand(...), default branch topk_metrics.max(-1, keepdim) &gt; eps expanded), a gt row is either fully valid or fully invalid, so the duplicates only ever land on index 0. stress-testing 30 random forward passes, count &gt; 1 appeared exclusively at column 0 and only on fully-invalid rows.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by raimbekovm** [comment]
> Thank you so much!
> […](#)

Happy to help!
